### PR TITLE
Enable model ONNX export

### PR DESCRIPTION
**Pull request by echarlaix**
Add an `onnx_export` function to enable the ONNX export of a model (removing the need to provide a HF model ID or path)

### CODE REVIEW

File: optimum/exporters/tasks.py [hunk 1]
```diff
@@ -1449,7 +1449,8 @@ def _infer_task_from_model_or_model_class(
             if auto_cls is None:
                 continue
             model_mapping = auto_cls._model_mapping._model_mapping
-            if target_name in model_mapping.values():
+            # Lowercase to avoid mismatch between MptForCausalLM and MPTForCausalLM
+            if target_name.lower() in (mapping.lower() for mapping in model_mapping.values()):
```
Comment (echarlaix) [inline optimum/exporters/tasks.py:1453]: can happen when `trust_remote_code=True`
https://huggingface.co/fxmarty/tiny-mpt-random-remote-code/blob/main/config.json#L4

File: optimum/exporters/tasks.py [hunk 2]
```diff
@@ -1552,6 +1553,18 @@ def infer_task_from_model(
 
         return task
 
+    @staticmethod
+    def _infer_library_from_model(model: Union["PreTrainedModel", "TFPreTrainedModel"]):
```
Comment (echarlaix) [inline optimum/exporters/tasks.py:1557]: was thinking about renaming `infer_library_from_model` to `infer_library_from_model_name_or_path` but don't want to introduce braking changes, an other solution would be to follow same pattern as https://github.com/huggingface/optimum/blob/36790792f65cbf9377c34c65617694c67006c105/optimum/exporters/tasks.py#L1519

File: optimum/exporters/onnx/__main__.py [hunk 3]
```diff
@@ -381,86 +392,160 @@ def main_export(
...
+    do_validation: bool = True,
+    model_kwargs: Optional[Dict[str, Any]] = None,
+    custom_onnx_configs: Optional[Dict[str, "OnnxConfig"]] = None,
+    fn_get_submodels: Optional[Callable] = None,
+    _variant: str = "default",
+    legacy: bool = False,
+    preprocessors: List = None,
+    device: str = "cpu",
+    no_dynamic_axes: bool = False,
+    task: Optional[str] = None,
```
Comment (echarlaix) [inline optimum/exporters/onnx/__main__.py:479]: I was initially planning for `onnx_export` to not have any `task` argument and to infer it given the model but there is one main issue : some models like `XLMWithLMHeadModel` belongs to multiple [`MAPPING_NAMES`](https://github.com/huggingface/transformers/blob/v4.37.0/src/transformers/models/auto/modeling_auto.py#L467)

File: optimum/exporters/onnx/convert.py [hunk 4]
```diff
@@ -95,7 +95,6 @@ def validate_models_outputs(
     onnx_files_subpaths: Optional[List[str]] = None,
     input_shapes: Optional[Dict] = None,
     device: str = "cpu",
-    dtype: Optional["torch.dtype"] = None,
```
Comment (fxmarty) [inline optimum/exporters/onnx/convert.py:98]: This one is apparently unused - thanks for fixing

File: tests/exporters/onnx/test_onnx_export.py [hunk 5]
```diff
@@ -586,3 +588,81 @@ def test_custom_export_trust_remote_error(self):
...
+                model_kwargs=model_kwargs,
+                device=device,
+                preprocessors=preprocessors,
+                task=task,
+            )
+
+    @parameterized.expand(_get_models_to_test(PYTORCH_EXPORT_MODELS_TINY))
+    @require_torch
+    @require_vision
+    def test_pytorch_export(
```
Comment (fxmarty) [inline tests/exporters/onnx/test_onnx_export.py:649]:
```suggestion
    @parameterized.expand(_get_models_to_test(PYTORCH_EXPORT_MODELS_TINY))
    @require_torch
    @require_vision
    @slow
    def test_pytorch_export(
```

File: optimum/exporters/onnx/__main__.py [hunk 6]
```diff
@@ -381,86 +392,160 @@ def main_export(
...
         not custom_architecture
-        and not is_stable_diffusion
+        and library_name != "diffusers"
         and task + "-with-past"
         in TasksManager.get_supported_tasks_for_model_type(model_type, "onnx", library_name=library_name)
     ):
-        if original_task == "auto":  # Make -with-past the default if --task was not explicitely specified
+        if (
+            original_task == "auto" and not monolith
+        ):  # Make -with-past the default if --task was not explicitely specified
```
Comment (fxmarty) [inline optimum/exporters/onnx/__main__.py:422]: Can you move the comment on its own line?

**Review by fxmarty — COMMENTED** [review]
Awesome!

Some issues I found:

```
(hf-inf) fxmarty@huggingface:~/hf_internship/optimum$ optimum-cli export onnx -m gpt2 --fp16 --device cuda gpt2_onnx
Framework not specified. Using pt to export the model.
Automatic task detection to text-generation (possible synonyms are: causal-lm).
```

should be `text-generation-with-past`

& then getting

```
Traceback (most recent call last):
  File "/home/fxmarty/hf_internship/optimum/optimum/exporters/onnx/__main__.py", line 674, in onnx_export
    use_subprocess=use_subprocess,
UnboundLocalError: local variable 'use_subprocess' referenced before assignment

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/fxmarty/anaconda3/envs/hf-inf/bin/optimum-cli", line 8, in <module>
    sys.exit(main())
  File "/home/fxmarty/hf_internship/optimum/optimum/commands/optimum_cli.py", line 163, in main
    service.run()
  File "/home/fxmarty/hf_internship/optimum/optimum/commands/export/onnx.py", line 249, in run
    main_export(
  File "/home/fxmarty/hf_internship/optimum/optimum/exporters/onnx/__main__.py", line 440, in main_export
    onnx_export(
  File "/home/fxmarty/hf_internship/optimum/optimum/exporters/onnx/__main__.py", line 689, in onnx_export
    raise Exception(
Exception: An error occured during validation, but the model was saved nonetheless at gpt2_onnx. Detailed error: local variable 'use_subprocess' referenced before assignment.
```

File: optimum/exporters/onnx/__main__.py [hunk 7]
```diff
@@ -381,86 +392,160 @@ def main_export(
...
             task = task + "-with-past"
         else:
             logger.info(
                 f"The task `{task}` was manually specified, and past key values will not be reused in the decoding."
                 f" if needed, please pass `--task {task}-with-past` to export using the past key values."
             )
+            model.config.use_cache = False
 
-    if task.endswith("-with-past") and monolith is True:
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/exporters/onnx/__main__.py:443]: If the task and monolith args are not compatible it makes sense to raise an error as soon as possible no ? (no need to load the model)

File: optimum/exporters/onnx/__main__.py [hunk 8]
```diff
@@ -381,86 +392,160 @@ def main_export(
...
+    legacy: bool = False,
+    preprocessors: List = None,
+    device: str = "cpu",
+    no_dynamic_axes: bool = False,
+    task: Optional[str] = None,
+    **kwargs_shapes,
+):
+    library_name = TasksManager._infer_library_from_model(model)
+    # framework = "pt" if is_torch_available() and isinstance(model, torch.nn.Module) else "tf"
+    dtype = model.dtype if library_name in {"transformers", "diffusers"} else model.config.torch_dtype
```
[the remark this answers is not in the document]
Comment (echarlaix) [reply optimum/exporters/onnx/__main__.py]: good point, the `model.config.torch_dtype` seems always set to `None`, replaced it in https://github.com/huggingface/optimum/pull/1649/commits/1a9999034b85d43574f361682bf3ba6d8a6392af

**Comment by echarlaix** [comment]
> Awesome!
> 
> Some issues I found:
> 
> ```
> (hf-inf) fxmarty@huggingface:~/hf_internship/optimum$ optimum-cli export onnx -m gpt2 --fp16 --device cuda gpt2_onnx
> Framework not specified. Using pt to export the model.
> Automatic task detection to text-generation (possible synonyms are: causal-lm).
> ```
> 
> should be `text-generation-with-past`
> 
> & then getting
> 
> ```
> Traceback (most recent call last):
>   File "/home/fxmarty/hf_internship/optimum/optimum/exporters/onnx/__main__.py", line 674, in onnx_export
>     use_subprocess=use_subprocess,
> UnboundLocalError: local variable 'use_subprocess' referenced before assignment
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/home/fxmarty/anaconda3/envs/hf-inf/bin/optimum-cli", line 8, in <module>
>     sys.exit(main())
>   File "/home/fxmarty/hf_internship/optimum/optimum/commands/optimum_cli.py", line 163, in main
>     service.run()
>   File "/home/fxmarty/hf_internship/optimum/optimum/commands/export/onnx.py", line 249, in run
>     main_export(
>   File "/home/fxmarty/hf_internship/optimum/optimum/exporters/onnx/__main__.py", line 440, in main_export
>     onnx_export(
>   File "/home/fxmarty/hf_internship/optimum/optimum/exporters/onnx/__main__.py", line 689, in onnx_export
>     raise Exception(
> Exception: An error occured during validation, but the model was saved nonetheless at gpt2_onnx. Detailed error: local variable 'use_subprocess' referenced before assignment.
> ```

thanks a lot @fxmarty, fixed it respectively in https://github.com/huggingface/optimum/pull/1649/commits/1ec4e0e8a59064d9f70b7975331e2a011a60ac93 and https://github.com/huggingface/optimum/pull/1649/commits/b27aacbba98f692e4872314f229f5b83ecc52784

**Review by fxmarty — APPROVED** [review]
LGTM thanks a lot!

File: optimum/exporters/onnx/__main__.py [hunk 9]
```diff
@@ -381,71 +380,43 @@ def main_export(
...
+    needs_pad_token_id = (
+        task == "text-classification"
+        and getattr(model.config, "pad_token_id", None)
+        and getattr(model.config, "is_decoder", False)
+    )
 
-    if custom_architecture and original_task == "auto":
-        raise ValueError(
-            f'Automatic task detection is not supported with custom architectures. Please specify the `task` argument. Suggestion: task="{task}" (or task="{task}-with-past" if the model is decoder-based and supports KV cache)'
-        )
```
Comment (fxmarty) [inline optimum/exporters/onnx/__main__.py:427]: Why was this removed?
